trestlebot/tasks/sync_upstreams_task.py

```diff
@@ -0,0 +1,168 @@
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""Trestle Bot Sync Upstreams Tasks"""
+
+import argparse
+import logging
+import pathlib
+import tempfile
+from typing import List, Optional
+
+from git import GitCommandError, Repo
+from trestle.common import file_utils
+from trestle.common.const import MODEL_DIR_LIST, VAL_MODE_ALL
+from trestle.common.err import TrestleError
+from trestle.common.model_utils import ModelUtils
+from trestle.core.base_model import OscalBaseModel
+from trestle.core.models.file_content_type import FileContentType
+from trestle.core.validator import Validator
+from trestle.core.validator_factory import validator_factory
+
+from trestlebot import const
+from trestlebot.tasks.base_task import ModelFilter, TaskBase, TaskException
+
+
+logger = logging.getLogger(__name__)
+
+
+class SyncUpstreamsTask(TaskBase):
+    """Sync OSCAL content from upstream git repositories."""
+
+    def __init__(
+        self,
+        working_dir: str,
+        git_sources: List[str],
+        model_filter: Optional[ModelFilter] = None,
+        validate: bool = True,
+    ) -> None:
+        """
+        Initialize sync upstreams task.
+
+        Args:
+            working_dir: Working directory to use for the task. Models from the sources will be copied
+            into this directory. It must be a valid trestle project root.
+            git_sources: List of upstream git sources to fetch from. Each source is a string
+            of the form <repo_url>@<ref> where ref is a git ref such as a tag or branch.
+            model_filter: Optional model filter to use for the task. This will filter models from
+            being copied from the upstream repositories.
+            validate: Optional argument to enable/disable validation of the models after they are copied
+
+        Notes: This task will fetch content from upstream repositories and copy it into the
+        trestle workspace. The task WILL overwrite any existing content in the workspace with the same
+        name. If it does not exist in the workspace, it will be created. Currently this only supports
+        OSCAL artifacts that are stored directly in the repository. This currently does not support
+        delete operations.
+        """
+        if not file_utils.is_valid_project_root(pathlib.Path(working_dir)):
+            raise TaskException(
+                f"Target workspace {working_dir} is not a valid trestle project root"
+            )
+        self.sources = git_sources
+        self.validate = validate
+        super().__init__(working_dir, model_filter)
```

Comment:
curious if model_filter has any applicability in execute()

Reply:
Good question. Since all tasks are used for workspace maintenance operations, the `model_filter` is used in a concrete [method](https://github.com/RedHatProductSecurity/trestle-bot/blob/5565f01ae982956365a9e621342a7403472567dd/trestlebot/tasks/base_task.py#L85) on the `TaskBase` class. In this task, it is called in `copy_validate_models` to allow certain paths to be skipped in the the clone repo during the copying process.